From 7a2e68731285f4cde2f130649d677a38fdaa6afe Mon Sep 17 00:00:00 2001
From: PratikBhusal <PratikBhusal@users.noreply.github.com>
Date: Mon, 24 May 2021 23:58:25 -0500
Subject: [PATCH] Add relative pixel border

This patch is the exact same as the st-relativeborder-0.8.3 patch by
Francesco Minnocci - ad17fmin (at) uwcad (dot) it

When working with a mixture of different DPI scales on different
monitors, you need to use a flexible font that will size correctly no
matter the DPI - for example, DejaVu Sans Mono-10. If you have a border
set in pixels, this border will look vastly different in size depending
on the DPI of your display.

This patch allows you to specify a border that is relative in size to
the width of a cell in the terminal.

Because it follows the "git --format-patch" file format, you may also
use "git am" to apply this patch.
---
 config.def.h | 4 +++-
 st.h         | 1 +
 x.c          | 2 ++
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/config.def.h b/config.def.h
index 6533be0..17fee21 100644
--- a/config.def.h
+++ b/config.def.h
@@ -4,9 +4,11 @@
  * appearance
  *
  * font: see http://freedesktop.org/software/fontconfig/fontconfig-user.html
+ * borderperc: percentage of cell width to use as a border
+ *		0 = no border, 100 = border width is same as cell width
  */
 static char *font = "Liberation Mono:pixelsize=12:antialias=true:autohint=true";
-static int borderpx = 2;
+static int borderperc = 20;

 /*
  * What program is execed by st depends of these precedence rules:
diff --git a/st.h b/st.h
index 8d26273..4d176d7 100644
--- a/st.h
+++ b/st.h
@@ -53,6 +53,7 @@ enum selection_snap {
 	SNAP_LINE = 2
 };

+int borderpx;
 typedef unsigned char uchar;
 typedef unsigned int uint;
 typedef unsigned long ulong;
diff --git a/x.c b/x.c
index 4f78bca..33d6bfe 100644
--- a/x.c
+++ b/x.c
@@ -1025,6 +1025,8 @@ xloadfonts(char *fontstr, double fontsize)
 	win.cw = ceilf(dc.font.width * cwscale);
 	win.ch = ceilf(dc.font.height * chscale);

+	borderpx = ceilf(((float)borderperc / 100) * win.cw);
+
 	FcPatternDel(pattern, FC_SLANT);
 	FcPatternAddInteger(pattern, FC_SLANT, FC_SLANT_ITALIC);
 	if (xloadfont(&dc.ifont, pattern))
--
2.31.1

